pipeline/secret.go

```diff
@@ -38,3 +47,158 @@ type (
 		Target string `json:"target,omitempty" yaml:"target,omitempty"`
 	}
 )
+
+var (
+	// ErrInvalidEngine defines the error type when the
+	// SecretEngine provided to the client is unsupported.
+	ErrInvalidEngine = errors.New("invalid secret engine")
+	// ErrInvalidOrg defines the error type when the
+	// org in key does not equal the name of the organization.
+	ErrInvalidOrg = errors.New("invalid organization in key")
+	// ErrInvalidRepo defines the error type when the
+	// repo in key does not equal the name of the repository.
+	ErrInvalidRepo = errors.New("invalid repository in key")
+	// ErrInvalidShared defines the error type when the
+	// org in key does not equal the name of the team.
+	ErrInvalidShared = errors.New("invalid team in key")
+	// ErrInvalidPath defines the error type when the
+	// path provided for a type (org, repo, shared) is invalid.
+	ErrInvalidPath = errors.New("invalid secret path")
+)
+
+// Purge removes the secrets that have a ruleset
+// that do not match the provided ruledata.
+func (s *SecretSlice) Purge(r *RuleData) *SecretSlice {
+	counter := 1
+	secrets := new(SecretSlice)
+
+	// iterate through each Secret in the pipeline
+	for _, secret := range *s {
+		if secret.Origin.Empty() {
+			// append the secret to the new slice of secrets
+			*secrets = append(*secrets, secret)
+
+			continue
+		}
+
+		// verify ruleset matches
+		if secret.Origin.Ruleset.Match(r) {
+			// overwrite the Container number with the Container counter
+			secret.Origin.Number = counter
+
+			// increment Container counter
+			counter = counter + 1
+
+			// append the secret to the new slice of secrets
+			*secrets = append(*secrets, secret)
+		}
+	}
+
+	return secrets
+}
+
+// ValidOrg returns true when the secret is valid for a given
+// organization.
+func (s *Secret) ValidOrg(org string) error {
```

Comment:
Functionality was moved from pkg-executor to types to allow cleaner validation on secrets